android/cameraview/cameraview/src/main/api21/com/google/android/cameraview/Camera2.java

```diff
@@ -170,23 +170,53 @@ public void onReady() {
         @Override
         public void onImageAvailable(ImageReader reader) {
             try (Image image = reader.acquireNextImage()) {
-                Image.Plane[] planes = image.getPlanes();
-                if (planes.length > 0) {
-                    ByteBuffer buffer = planes[0].getBuffer();
+                if (image.getFormat() == ImageFormat.JPEG) {
+                    ByteBuffer buffer = image.getPlanes()[0].getBuffer();
                     byte[] data = new byte[buffer.remaining()];
                     buffer.get(data);
-                    if (image.getFormat() == ImageFormat.JPEG) {
-                        mCallback.onPictureTaken(data);
-                    } else {
-                        mCallback.onFramePreview(data, image.getWidth(), image.getHeight(), mDisplayOrientation);
-                    }
-                    image.close();
+
+                    mCallback.onPictureTaken(data);
+                } else { // ImageFormat.YUV_420_888
+                    byte[] data = YUV_420_888toNV21(image);
+
+                    int w = image.getWidth();
+                    int h = image.getHeight();
+
+                    // this part fixes an issue when YUV to NV21 conversion doesn't work (Pixel 2)
+                    // properly so the result image is skewed and barcodes can't be detected
+                    // it appears there is a padding must be considered
+                    Image.Plane plane = image.getPlanes()[0];
+                    int ps = plane.getPixelStride();
+                    int rs = plane.getRowStride();
+                    int padding = rs - ps * w;
+                    int dw = padding / ps;
+
+                    mCallback.onFramePreview(data, w + dw, h, mDisplayOrientation);
                 }
             }
         }
 
     };
 
+    // WARNING
+    // this is a simplified version of a conversion, but it's enough for QR code detection
+    // for a proper conversion U-channel should be extracted into byte array result
```

Comment:
can we do a proper conversion and include the U-channel? I'm not aware of any reason not to and it seems quite simple: https://stackoverflow.com/questions/40090681/android-camera2-api-yuv-420-888-to-jpeg

Reply:
Sure.